### ISSUE TITLE
Settings should have a way to express "no opinion" for a setting

### ISSUE THREAD

**Issue by leifwalsh** [github/safe-settings#412]
The current behavior of settings.yml is that a lot of fields are required. They can be `null`, or they can be filled out, but they must be present (otherwise the API returns a 422). This means you can either specify things, or turn them off, but you cannot express the concept "safe-settings has no opinion on this setting, and each repository is free to make its own choices".

I'll describe the example I just encountered:
* I have an organization with several repos, and each repo has some set of checks that I want to make required for PRs to land
* But the set of checks is different depending on the repo
* With safe-settings, I can specify `branches.protection.required_status_checks` but I can't provide one coherent value for `contexts`, since the check names are different across repos.
* I could specify `contexts: []` but this prevents me from _adding_ contexts within one repo in the org: safe-settings removes my required check.
* I also tried specifying `branches.protection.required_status_checks: null`, hoping that this would just leave the setting alone. Instead, it disabled `required_status_checks`, even if I enabled it for one repo.

It would be nice for some of the fields to be optional (meaning safe-settings will just leave them alone), or to have values that are "minimums". In the case of `required_status_checks`, I might want to say that every repo needs to enforce that the linter check passes, or the secret scanning check passes (though, with [Required Workflows](https://docs.github.com/en/actions/using-workflows/required-workflows) this will be taken care of without safe-settings), but allow repos to add other checks they require, on top of that.

**Comment by leifwalsh** [comment]
Before you ask, yes I'm aware of suborgs and per-repo settings. Using those to specify required checks individually per repo would require that the repo owners coordinate with the org owners to add required checks. I'd rather have things set up so that they can act independently, self-serving their settings right in the github UI, whenever they want to change required checks (or modify any other settings the whole org doesn't need to be opinionated about).

**Comment by leifwalsh** [comment]
I poked around a little bit, and this change allows you to omit `required_status_checks` from your `settings.yml` and avoid changing the checks: https://github.com/github/safe-settings/compare/main-enterprise...leifwalsh:safe-settings:patch-1

However, I'm not proposing merging it because I don't think it's sufficiently generic, and someone who knows this codebase better could probably do it in a way that works for any setting.

**Comment by martinm82** [comment]
Hi @leifwalsh,

I have exactly the same issue. I am trying to introduce safe-settings on an organization with thousands of repositories. Each of those repos already have some branch protection rules (e.g. status checks) defined. 
I am incrementally introducing settings to the organization and basic branch protection rules are the next ones but I am struggling with rolling this out (without changes in the code) as it will override all more stricter protection rules.

One solution would be to automatically fetch all the branch protection rules through the REST API and convert them into repo settings files and store them in `.github/repos`. But we are talking here about 3.5k files and honestly I do not want to do that now as my goal is to teach developers to define suborg configuration files instead.

Regarding your solution: I guess having a feature flag in for example the deployment configuration would be a good step to opt-in for this kind of behaviour. But would be good to know what the others think about this @decyjphr @anderssonjohan @svg153 ?

**Comment by decyjphr** [comment]
> I am incrementally introducing settings to the organization and basic branch protection rules are the next ones but I am struggling with rolling this out (without changes in the code) as it will override all more stricter protection rules.

We do have the [`override validators` ](https://github.com/github/safe-settings#custom-rules) which can prevent overriding more stricter rules with the default ones. This happens at the plugin level, so it is all or nothing for that plugin. i.e. the entire `branch protection rule` is allowed or skipped based on the validation result, and you cannot just skip applying a subsection of the branch protection rule,

> One solution would be to automatically fetch all the branch protection rules through the REST API and convert them into repo settings files and store them in `.github/repos`. But we are talking here about 3.5k files and honestly I do not want to do that now as my goal is to teach developers to define suborg configuration files instead.

__Option: Flag to indicate if a `github` settings should be merged on top of the consolidated settings__
Currently, before applying the changes, we get the `github` settings in order to use it in  `mergeDeep.comparedeep`, so we do have the settings from GitHub for each plugin. Maybe if there is a flag like, `merge_strategy: [combine-github-plugin | combine-plugin-github | overwrite}`, to indicate that `github` settings should be `combined`, with`plugin` or `github` taking precedence, or overwrite `github`.  Also, we can have an option that if run in `pull request` mode that these changes are persisted in a `repo-level` settings file. This is not as bad since the files would only be in the PR branch.

This would be done at the plugin level (to keep it simple) and we can merge the entire setting from `github` with the `consolidated` settings for the plugin. I think with this change we would be able to keep `contexts` from the existing branch protection while applying other changes. 

A more complex implementation would be to do this at a more granular level and I am not sure how we we would implement that.

**Comment by EricBartusch** [comment]
> I poked around a little bit, and this change allows you to omit `required_status_checks` from your `settings.yml` and avoid changing the checks: [main-enterprise...leifwalsh:safe-settings:patch-1](https://github.com/github/safe-settings/compare/main-enterprise...leifwalsh:safe-settings:patch-1)
> 
> However, I'm not proposing merging it because I don't think it's sufficiently generic, and someone who knows this codebase better could probably do it in a way that works for any setting.

I took this and got it working for us. I have `required_status_checks` commented out in our admin repo and it completely ignores the section without error now:
```                
if (params.required_status_checks === undefined) {
    const { strict, checks } = result.data.required_status_checks
    params.required_status_checks = { strict, checks }
}
```

**Comment by leifwalsh** [comment]
Yeah, that code works for a specific use case but doesn't generalize well.

**Comment by n-oden** [comment]
At the risk of necro-ing a now two year old bug report— has there been any forward progress on this?

I'm in exactly the same position as @leifwalsh and @martinm82 here, and I can't possibly be the only one: I'm trying to retrofit safe-settings onto a large installed base of existing repositories, each with their own hand-tooled list of required_status_checks, and for now it looks like using a forked copy of safe-settings with @leifwalsh's patch applied will be the path of least resistance/surprise, but ugh.

My own opinion: I don't necessarily need safe-settings to have some kind of intelligent auto-merge logic here. Just extending the pattern of "a `null` or empty object `{}` means leave the existing resource alone" which already exists in a few places would be sufficient.
